sdk/src/test/java/com/qonversion/android/sdk/internal/user/controller/UserControllerTest.kt

```diff
@@ -0,0 +1,206 @@
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Nested
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertDoesNotThrow
+
+internal class UserControllerTest {
+
+    private lateinit var userController: UserControllerImpl
+
+    private val mockUserService = mockk<UserService>()
+    private val mockUserCacher = mockk<Cacher<User?>>()
+    private val mockLogger = mockk<Logger>()
+
+    private val slotInfoLogMessage = slot<String>()
+    private val slotErrorLogMessage = slot<String>()
+    private val slotVerboseLogMessage = slot<String>()
+    private val mockUser = mockk<User>()
+
+    @BeforeEach
+    fun setUp() {
+        userController = UserControllerImpl(mockUserService, mockUserCacher, mockLogger)
+
+        every { mockLogger.info(capture(slotInfoLogMessage)) } just runs
+        every { mockLogger.error(capture(slotErrorLogMessage)) } just runs
+        every { mockLogger.error(capture(slotErrorLogMessage), any()) } just runs
+        every { mockLogger.verbose(capture(slotVerboseLogMessage)) } just runs
+    }
+
+    @Nested
+    @ExperimentalCoroutinesApi
+    inner class GetUserTest {
+        private val userId = "userId"
+
+        @BeforeEach
+        fun setUp() {
+            every {
+                mockUserService.obtainUserId()
+            } returns userId
+        }
+
+        @Test
+        fun `get user from cache with default state`() = runTest {
+            // given
+            every {
+                mockUserCacher.getActual()
+            } returns mockUser
+
+            // when
+            val result = userController.getUser()
+
+            // then
+            assertThat(result).isEqualTo(mockUser)
+            assertThat(slotVerboseLogMessage.captured).isEqualTo("getUser() -> started")
+        }
+
+        @Test
+        fun `get user from API successfully`() = runTest {
+            // given
+            userController = spyk(userController)
+
+            every {
+                mockUserCacher.getActual()
+            } returns null
+
+            coEvery {
+                mockUserService.getUser(userId)
+            } returns mockUser
+
+            every {
+                userController.storeUser(mockUser)
+            } just runs
+
+            // when
+            val result = userController.getUser()
+
+            // then
+            assertThat(result).isEqualTo(mockUser)
+            assertThat(slotInfoLogMessage.captured).isEqualTo("User info was successfully received from API")
+            coVerifyOrder {
+                mockUserCacher.getActual()
+                mockUserService.obtainUserId()
+                mockUserService.getUser(userId)
+                userController.storeUser(mockUser)
+            }
+        }
+
+        @Test
+        fun `get user from cache with error state`() = runTest {
+            // given
+            userController = spyk(userController)
```

Comment:
no need in spyk here

Reply:
I use it to mock the `storeUser` function